### ISSUE TITLE
[Error: Cannot find module '../build/Release/bson'] code: 'MODULE_NOT_FOUND'

### ISSUE THREAD

**Issue by djtregear** [Automattic/mongoose#3436]
Receiving the following error message:

{ [Error: Cannot find module '../build/Release/bson'] code: 'MODULE_NOT_FOUND' }
js-bson: Failed to load c++ bson extension, using pure JS version

This occurs on both development (OSX) and live environments (Ubuntu 14.04) when using Node 4.1.2 (which is installed using 'n') and with a clean install of 'node_modules'.

If I roll back to Node 0.12.7 and reinstall 'node_modules' everything works fine within both environments.

Would really like to get this resolved and am happy to work through this as long as someone gives me a lead.

Thanks in advance

**Comment by vkarpov15** [comment]
which version of mongoose? Also, it's not an error, its a warning.

**Comment by djtregear** [comment]
I upgraded today to mongoose v4.1.10, to see if this warning was resolved, but the same situation was occurring a week or so ago against Node v4.0.0/Node v0.12.7 and mongoose v3 (latest version).

If it's a warning I assume that the lack of binary bson implementation would impact production performance?

Thanks for looking into this

**Comment by YouriT** [comment]
It seems that mongoose relies on mongondb@2.0.43 which relies on on bson@0.4.11 which is not available in the NPM registry anymore. Don't know why, but I ran into the same error as you @djtregear today...

```$ npm view bson versions
...
  '0.4.10',
  '0.4.12',
  '0.4.13',
  '0.4.14',
  '0.4.15' ]
```

No 0.4.11 as you can see. It has already been reported as you can see here mongodb/js-bson#147

So a new release should be done on the mongoose side I guess.

**Comment by YouriT** [comment]
@djtregear maybe you are using hex7c0/mongodb-backup as I do, which has a hard dep on bson@0.4.11 and causes the crash in my case.

**Comment by vkarpov15** [comment]
Not necessarily, the new JS parser/serializer should be pretty close in terms of performance and memory usage.

Also, define what you mean by "latest mongoose v3". Please don't use 3.9.7, its deprecated in npm and gives you loud warnings when you try to use it for a reason. Use `~3.8.x` if you want to use mongoose 3.

**Comment by cleechtech** [comment]
I am getting the same error with 

```
➜  express-auth git:(master) node -v
v4.2.1
➜  express-auth git:(master) npm -v
3.3.6
```

I have mongoose 3.8.1 in my package.json. And the error:

```
ode server
{ [Error: Cannot find module '../build/Release/bson'] code: 'MODULE_NOT_FOUND' }
js-bson: Failed to load c++ bson extension, using pure JS version
module.js:339
    throw err;
    ^

Error: Cannot find module './auth'
    at Function.Module._resolveFilename (module.js:337:15)
    at Function.Module._load (module.js:287:25)
    at Module.require (module.js:366:17)
    at require (module.js:385:17)
    at Object.<anonymous> (/Users/cleech/Projects/express-auth/config/passport.js:7:18)
    at Module._compile (module.js:435:26)
    at Object.Module._extensions..js (module.js:442:10)
    at Module.load (module.js:356:32)
    at Function.Module._load (module.js:311:12)
    at Module.require (module.js:366:17)
    at require (module.js:385:17)
    at Object.<anonymous> (/Users/cleech/Projects/express-auth/server.js:19:1)
    at Module._compile (module.js:435:26)
    at Object.Module._extensions..js (module.js:442:10)
    at Module.load (module.js:356:32)
    at Function.Module._load (module.js:311:12)
```

**Comment by vkarpov15** [comment]
The 'MODULE_NOT_FOUND' is just a warning. The error is in express-auth according to the stack trace.

**Comment by cleechtech** [comment]
Ok I'm an idiot. Thank you. I appreciate the mongoose project. It is awesome

**Comment by KO527** [comment]
Here is the problem. I had the issue found at Error: Cannot find module 'mongodb/node_modules/bson' #2718 and then I changed mongoose. 3.8.23 to 4.2.6. After that I got the error 
{ [Error: Cannot find module '../build/Release/bson'] code: 'MODULE_NOT_FOUND' }
js-bson: Failed to load c++ bson extension, using pure JS version

And @vkarpov15, In reference to what @djtregear said, having this error does affect performance on production. Works perfectly fine locally, but during production, things go awfully wrong with this error.

**Comment by KO527** [comment]
i forgot to post the rest of the log....This is how it looks if it even means anything really


MEAN.JS application started on port 8080
Now using node v5.5.0
Found package.json: ./package.json
Start script specified in package.json: grunt
Running command: npm start

> dayniteevents@0.0.1 start /mnt/app
> grunt

Running "jshint:all" (jshint) task

[✓] Logs successfully retrieved.

**Comment by vkarpov15** [comment]
Try clearing out your `node_modules` and re-running npm install. `rm -rf node_modules && npm install`.

**Comment by KO527** [comment]
Nope, that did not work. Modulus logs gives me the same exact error.

**Comment by KO527** [comment]
Did anybody on your team come up with anything else or figure out what may be wrong with the app’s synchronization with Modulus.io?

**Comment by vkarpov15** [comment]
You might also want to try upgrading to latest mongoose, that warning message isn't even in the latest version of mongoose. Also, are you using a separate version of the mongodb driver somewhere? Try running `npm list | grep mongo` and show me the output

**Comment by KO527** [comment]
Thank you for responding but I actually talked to someone at the modulus.io support team and we were able to figure out what was wrong, but i am now running mongoose 4.4.5 and the output i get when I run those commands are:

├── connect-mongo@1.1.0
├─┬ mongodb@2.1.8
│ ├─┬ mongodb-core@1.3.5
├─┬ mongoose@4.4.7
│ ├─┬ mongodb@2.1.7
│ │ ├─┬ mongodb-core@1.3.1

Apparently I have things configured for the production stage in my working directory that is meant to be left up to the modulus platform to take care and thats why I had a conflict. I have yet to resolve this on my  own because I am working on different projects but thank you for your input

**Comment by vkarpov15** [comment]
Great, glad you figured it out. FWIW I always recommend people stop using connect-mongo, I wrote [connect-mongodb-session](https://www.npmjs.com/package/connect-mongodb-session) because connect-mongo is really quirky.

**Comment by KO527** [comment]
Thanks for the advice on that. I definitely will check that out.

**Comment by simxas** [comment]
I fixed this problem by upgrading mongoose and using this command `rm -rf node_modules && npm install`

**Comment by coyle** [comment]
fwiw I fixed this by upgrading my global npm dependencies npm update -g

**Comment by vkarpov15** [comment]
Installing mongoose with -g hurts my soul :(

**Comment by suresh-mobilocity** [comment]
I am having the same problem js-bson: Failed to load c++ bson extension, using pure JS version. I node-gyp failed to compile and link. 
if ignore it as warning my application is not able to insert any data into database.
calling to insert document into db. I am using 

error found in insert message TypeError: Argument must be a string
    at TypeError (native)
    at Buffer.write (buffer.js:761:21)
    :
Connection.write (/home/snapsolv/socketio/node_modules/mongodb/lib/mongodb/connection/connection.js:265:37

 as my application is using 

   var msgObjectID = mongoose.Types.ObjectId();
 "_id":msgObjectID,


node -v  -- v6.8.1

MongoDB 2.6.9
Ubuntu 14.04
mongoose 4.0.1

The same bson extension build works fine in another environment with older version of node 5.11 and Ubuntu 16.04

If Update my dependency packages mongoose or mongodb or bson my code breaks..

**Comment by Dommagg** [comment]
I wrote this code
```
var Server = require("mongo-sync").Server;
var server = new Server('127.0.0.1');
// var result = server.db("Room").getCollection("restaurants").find().count();

var Fiber = require('fibers');
Fiber(function() {
	// set up goes here ...
	function called()
	{
		var result = server.db("Room").getCollection("restaurants").find().toArray();
		console.log('length: ', result.length);
		server.close();
		return 'Success!';
	}
	var status = called();
	console.log(status, 'hi');
}).run();
```
It has executed and showed output as follows

{ Error: Cannot find module '../build/Release/bson'
    at Function.Module._resolveFilename (module.js:469:15)
    at Function.Module._load (module.js:417:25)
    at Module.require (module.js:497:17)
    at require (internal/module.js:20:19)
    at Object.<anonymous> (/var/www/Room51/node_modules/bson/ext/index.js:15:10)
    at Module._compile (module.js:570:32)
    at Object.Module._extensions..js (module.js:579:10)
    at Module.load (module.js:487:32)
    at tryModuleLoad (module.js:446:12)
    at Function.Module._load (module.js:438:3) code: 'MODULE_NOT_FOUND' }
js-bson: Failed to load c++ bson extension, using pure JS version
length: 2842
Success! hi

I need a solution for   **Cannot find module '../build/Release/bson'** 
why this issue is raised?

**Comment by kartset** [comment]
I fixed this problem by upgrading mongoose to 4.4.5. as @KO527 mentioned. Thank you guys.
